### PR TITLE
 #6 Snackbar and SnackbarNotification components implemented

### PR DESCRIPTION
**Pull request by luizmendez**
Implemented `<Snackbar>` and `<SnackbarNotification>` components with tests and added their documentation to Storybook.

### CODE REVIEW

File: _stories/molecules/Snackbar/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+## Snackbar
+
+The `<Snackbar>` component is a list of notifications that the user can pass as children using the `<SnackbarNotification>` component.
```
[the remark this answers is not in the document]
Comment (luizmendez) [reply _stories/molecules/Snackbar/README.md]: Thanks for the observation, fixed.

File: _stories/molecules/Snackbar/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+## Snackbar Notification
+
+The `<SnackbarNotification>` component displays a notification inside the `<Snackbar>` component wrapper.
+The notification can be hidden by clicking a button or waiting a timeout.
```
[the remark this answers is not in the document]
Comment (luizmendez) [reply _stories/molecules/Snackbar/README.md]: Thanks, fixed.

File: _stories/molecules/Snackbar/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+A callback function can be passed to the component to be called after it hides,
```
[the remark this answers is not in the document]
Comment (luizmendez) [reply _stories/molecules/Snackbar/README.md]: Fixed.

File: _stories/molecules/Snackbar/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+| --------------- | -------------------- | --------------------------------------------------------------------------------------------------------------------- | ------- |
+| id              | `String` `Number` | Unique key for identification                                                                                         |         |
+| text            | `String`             | Content of the notification, can be passed as children. If no text or children is passed, the component won't render. |         |
+| context         | `String`             | Context of notification (success, info, warning, danger, hidden)                                                      | ``      |
+| buttonContext   | `String`             | Context of close button, if not passed uses context value (succes, info, warning, danger, dark)                       | ``      |
+| msToClose       | `Number`             | Time in ms to hide notification, if 0 is passed ignores timeout                                                       | `3000`  |
```
[the remark this answers is not in the document]
Comment (luizmendez) [reply _stories/molecules/Snackbar/README.md]: Fixed.

File: components/atoms/SnackbarNotification.jsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        // Sets timeout in state
+        if (timeout) {
+            this.setState({
+                timeout
+            });
+        }
+    };
+
+    _clearTimeout = () => {
+        if (this.state.timeout) clearTimeout(this.state.timeout);
```
[the remark this answers is not in the document]
Comment (luizmendez) [reply components/atoms/SnackbarNotification.jsx]: I think it's better to keep this in a property, was setting it on state to keep the timeout on rerenders but it's the same.

File: components/atoms/SnackbarNotification.jsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    };
+
+    // Sets the hide in state and calls callback function
+    _closeSnackbar = () => {
+        // Clear timeout if still running
+        this._clearTimeout();
+        // If there's a callback function pass the id
+        const { callback, id } = this.props;
+        const cb = callback ? (id ? () => callback(this.props.id) : callback) : null;
```
[the remark this answers is not in the document]
Comment (luizmendez) [reply components/atoms/SnackbarNotification.jsx]: Waiting for confirmation if the id would be necessary, another solution is creating a random id on mount if one is not passed by props.

**Comment by luizmendez** [comment]
@edrpls Thanks for the review, I left notes on fixed issues, I think I'm only waiting the comments on the props.id issue.

Thanks again.

File: _stories/molecules/Snackbar/index.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    success: 'success',
+    info: 'info',
+    warning: 'warning',
+    danger: 'danger',
+    'No Value': ''
+};
+
+const component = () => (
+    <Snackbar>
+        <SnackbarNotification
```
[the remark this answers is not in the document]
Comment (jscottsmith) [reply _stories/molecules/Snackbar/index.js]: I think it needed a `displayName` property, and it looks like it was added.

File: _stories/molecules/Snackbar/README.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
+## Snackbar
+
+The `<Snackbar>` component is a list of notifications that the user can pass as children using the `<SnackbarNotification>` component.
+
+## Snackbar Notification
+
+The `<SnackbarNotification>` component displays a notification inside the `<Snackbar>` component wrapper.
+The notification can be hidden by clicking a button or waiting a timeout.
+A callback function can be passed to the component to be called after it hides, 
+the first parameter of the callback will be the id of the notification if it is passed as a prop.
```
[the remark this answers is not in the document]
Comment (jscottsmith) [reply _stories/molecules/Snackbar/README.md]: Ya, I don't think anything else needs to be passed in the callback.

File: components/atoms/SnackbarNotification.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (jscottsmith) [reply components/atoms/SnackbarNotification.jsx]: I think for now just requiring the prop `id` will be good -- all our current uses manage the ids externally.

File: components/atoms/SnackbarNotification.jsx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        /** Styles passed through props */
+        style: PropTypes.object,
+        /** Class names passed through props */
+        className: PropTypes.string
+    };
+
+    static defaultProps = {
+        msToClose: 3000,
+        hideCloseButton: false,
+        style: {},
```
Comment (jscottsmith) [inline components/atoms/SnackbarNotification.jsx]: I would prefer to just remove the `style` and `className` defaults, having `undefined` shouldn't break anything and makes snapshots cleaner

File: components/molecules/Snackbar.jsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    /** Class names passed through props */
+    className: PropTypes.string,
+    /** Styles passed through props */
+    style: PropTypes.object,
+    /** Component children to use as notifications, use <SnackbarNotification> component */
+    children: PropTypes.node
+};
+
+Snackbar.defaultProps = {
+    className: '',
```
Comment (jscottsmith) [inline components/molecules/Snackbar.jsx]: same note: to remove `style` and `className` defaults

File: components/atoms/SnackbarNotification.jsx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+class SnackbarNotification extends Component {
+    static propTypes = {
+        /** Unique key for identification */
+        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+        /** Message or content of the notification */
+        text: PropTypes.string,
+        /** Context of notification (success, info, warning, danger, hidden) */
+        context: PropTypes.string,
+        /** Context of close button, if not passed uses context value (succes, info, warning, danger, dark) */
```
Comment (jscottsmith) [inline components/atoms/SnackbarNotification.jsx]: minor:  `succes` -> `success`

File: _stories/molecules/Snackbar/README.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+### Props
+
+The following props may be passed to configure the `<SnackbarNotification>` component:
+
+| name            | type                 | description                                                                                                           | default |
+| --------------- | -------------------- | --------------------------------------------------------------------------------------------------------------------- | ------- |
+| id              | `String` `Number` | Unique key for identification                                                                                         |         |
+| text            | `String`             | Content of the notification, can be passed as children. If no text or children is passed, the component won't render. |         |
+| context         | `String`             | Context of notification (success, info, warning, danger, hidden)                                                      | ``      |
+| buttonContext   | `String`             | Context of close button, if not passed uses context value (succes, info, warning, danger, dark)                       | ``      |
```
[the remark this answers is not in the document]
Comment (mnicole) [reply _stories/molecules/Snackbar/README.md]: i'm wondering if it should be something not exposed to the user, so the component would automatically set it based on the context value. i'm not sure if someone would need to override it manually with another class.

**Comment by mnicole** [comment]
@luizmendez thanks so much for contributing! this looks great and super thorough 💯

File: _stories/molecules/Snackbar/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (edrpls) [reply _stories/molecules/Snackbar/README.md]: I don't see the fix.

File: _stories/molecules/Snackbar/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (edrpls) [reply _stories/molecules/Snackbar/README.md]: dunno, I guess it'd be good to have defaults, but also leave that possibility open for the user, they don't necessarily have to use it.

File: _stories/molecules/Snackbar/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (edrpls) [reply _stories/molecules/Snackbar/README.md]: don't see this either

File: _stories/molecules/Snackbar/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (luizmendez) [reply _stories/molecules/Snackbar/README.md]: Thanks, gonna leave it like that.

File: _stories/molecules/Snackbar/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (luizmendez) [reply _stories/molecules/Snackbar/README.md]: I think it's a good idea, it's there if by any chance is needed but it's not disclosed to the user as the natural use would be expecting to use the same context in both.

File: components/atoms/SnackbarNotification.jsx (hunk 9 above)
Comment (luizmendez) [reply components/atoms/SnackbarNotification.jsx]: Thanks, fixed!

File: components/molecules/Snackbar.jsx (hunk 10 above)
Comment (luizmendez) [reply components/molecules/Snackbar.jsx]: Fixed.

**Comment by luizmendez** [comment]
@edrpls @mnicole @jscottsmith Thanks again for the review, all issues should be fixed in the last push.

**Pull request merged**